### DISCUSSION TITLE
RS485 8 Relay Board

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#497] [REM]
@rstrouse I have a requirement for a remote relay board that I would like to integrate with REM. My valve wires were not long enough to reach the main control panel.  I currently have a "[DIN Rail Mount 8 SPDT Relay](https://www.amazon.com/Electronics-Salon-Interface-Module-Operating-Voltage/dp/B00M1MC3ZU)" that I leverage with my own solution, but as you know I'm trying to move to REM.

To that end, I'm interested in preferably buying but not opposed to building a similar module that can be controlled remotely via REM.  

Have you looked at integrating any existing RS485 8 relay boards?

**Comment by rstrouse** [comment]
That board can be driven via gpio or using an MCP or PFC type gpio extender already in REM.  I believe @drjeff is using one.

**Reply by DrJeff** [reply]
I was using one and it works great via i2C or GPIO, I have since installed controller outside and and connected to BAS and 8 relay boards. How short are your valve wires?

**Reply by orrious** [reply]
The issue with using a MCP or PFC is they will generally use I2C or SPI, both of which aren't intended to go beyond 1 meter.  RS485 would be my preferred transport as it's reliable and intended to traverse longer distances (4000 feet).  

Just searching around I came across this "[RS485 Modbus Relay Delay Switch DC 12V 8-Channel Relay](https://www.amazon.com/Modbus-8-Channel-Control-9600bps-Monitor/dp/B09QS211LH)"  Any interest in supporting RS485 Modbus in REM?

**Reply by tagyoureit** [reply]
rstrouse found this awesome i2c extender (it can go up to 3m).  We are both using it in our standard builds as it improves signal quality even in shorter distances.  And it's only $7.50.

https://www.adafruit.com/product/4756

**Reply by orrious** [reply]
The valve wires were full length, but still would not reach where my controller was.  They have long been terminated in their own enclosure.  The "valve relay" enclosure is about 50 foot away from the main control panel, as the wire runs through wire management, etc. So even an I2C extender won't reliably work.  Unfortunately, I2C / SPI at this distance is outside of what it was designed for.

Is the RS485 protocol that Pentair uses Modbus or something else?  Just wondering how difficult it would be to get the above board working vs reflash that board with custom firmware.

**Reply by rstrouse** [reply]
The pentair RS485 is not modbus.  In fact there are a couple of different protocols on the bus.  In fact simply adding modbus slaves on the bus would likely confuse all the other slaves on the bus.  However a tcp socket interface or multiple rs485 buses would be doable.  Do you have the pi located off the pad?

**Reply by DrJeff** [reply]
@orrious Are you using Home Assistant? Are the only thing you trying to control separately are the Valves?

**Reply by orrious** [reply]
@rstrouse the pi is located on the pad.  I'll put together a diagram of how it is now.  To be fair, I'm already using 5v GPIOs from a mega in the main control panel to drive the relay board, so there is nothing stopping me from continuing to do so with a different 5v GPIO source.  It's a bit a of a pain due to needing to source 10 conductor wire to go between the main control panel and relay.  Running RS485 could make remote relays more practical for other folks.

**Reply by orrious** [reply]
@DrJeff I do run HA, NodeRed, and two dozen other containers.  I've been running a custom solution leveraging it all for over 2 years, but would like to standardize on REM with no dependency on HA.  My pi is currently connected via Wifi and even with an AP very near, it looses connectivity due to RF interference.  I.e network is not reliable.  I've even had to go so far as to run a local mosquito instance on the pi to increase stability.  It is my desire to remove dependency on the network and external services.

**Reply by DrJeff** [reply]
Ok I get it that's what I did removed all of HA and Node Red from pool (The logic, well mostly my lights are inside of esphome since I use 1 Large transformer 1200W for pool lights and yard lights), But that entailed adding the Pi at the equipment. So if you have the Pi at the equipment it's unclear how your valves can't be connected to your Pi if they were cut they can be extended I cut those all the time and have added more cable as well. Yes do show your layout

**Reply by orrious** [reply]
Drawing it up as we speak.  Keep in mind, the pi's GPIOs, IMO are not usable.  When the pi reboots, the GPIOs change state and can accidentally dead head your system, etc.  I need a minimum of 20 GPIO 5V output.  6 analog input, 2 pulse counters.  I'm currently using an arduino mega via I2C to the Pi today.  Can you recommend a different board(s) that could provide the same?

**Reply by rstrouse** [reply]
Maybe having a better handle on your layout may spark some ideas.  In the meantime there are a couple of things to note.  First, you may have as many instances of REM spread throughout the world and njsPC will talk to each one at the same time.  Second, an RS485 relay is not out of the realm of possibility.  njsPC can also speak to many RS485 comm ports at the same time.  I looked for some RS485 relays and didn't find much that had any real documentation though.

**Reply by orrious** [reply]
The board I provided above uses standard MODBUS protocol.  Here are the libraries https://www.npmjs.com/package/modbus-serial.  I would agree that I would probably run this on a separate RS-485 bus.  Let me know if this is something you want to tinker with, I'll happily donate a relay board to the cause.

Not sure if you saw this:
![710Wz7P6fcL _SX522_](https://user-images.githubusercontent.com/9359564/175665671-e8680001-a0a9-41f3-a555-1bb2fed2b209.jpg)

**Comment by rock-crusher** [comment]
I may not fully understand your entire dilemma here, but couldn’t you simply use REM together with the Sequent 8 relay hat in order to drive your existing relay board for the valves using the existing 10 conductor cable?
Sorry if I’m misunderstanding what you’re trying to accomplish.

**Reply by orrious** [reply]
I would not drive one relay with another.  If I were to drive the existing relay board, it would be with 5v GPIO, like I am today.  Understand that when I design a solution, it's not just for myself.  That 10 conductor, stranded, tinned, in a UV resistant jacket was a pain to source and wasn't cheap.  I will also need 8x 5v GPIOs to drive it.  If I can do the exact same thing with 2 conductors, RS485, and no GPIOs, why wouldn't I?  

From a price perspective, the relay board I have is $50, $40 for the cable + Pi Hat with 8 GPIOs vs the $29 RS485 8 relay board I provided above.

**Reply by rstrouse** [reply]
Back when Raspberry Pi wasn't made of unicorn farts I would have suggested you just pop another pi in your relay box.  njsPC has no limit to the number of distributed REM interfaces it will talk to.  Alas though I am pretty sure if you find one it will smell like skittles and buy like caviar.

**Reply by orrious** [reply]
Rpi are being reserved for industry first and then the scalpers pick up anything left.  Its ridiculous.  

Can you suggest a hat or multiple hats that can provide 20 GPIO 5V output, 6 analog input, 2 pulse counters, and RS485?

Have you talked to Sequent Microsystems about developing a board for us, rather than getting a lot what we don't need with some of their other boards.  I like their concept of edge connectors and stackable, but I'm trying to find something that has a 20+ outputs.

**Reply by rstrouse** [reply]
Perhaps the most common I/O extender is the MCP2017.  That being said it is I2c controlled and given the layout of your pad I am pretty sure an LCT4311 chip can go the distance to an MCP2017 relay board.  They claim 100 feet at the I2c speeds we are talking about here.

**Reply by rock-crusher** [reply]
`@orrious:  I would not drive one relay with another   `

It’s done all the time, Inquiring minds want to know why not?

**Reply by tagyoureit** [reply]
> That 10 conductor, stranded, tinned, in a UV resistant jacket was a pain to source and wasn't cheap.

$40 for that cable doesn't sound too bad!  :)

We moved away from chips with the Songle/Sainsmart relays because they are a low level trigger.  As one really smart person (cough, @rstrouse) told me... 

>  With a low level trigger here is what happens.  When the pi is powered up there is a short amount of time where all the pins are low while it is booting up and if there is power on the relay board it will be in a triggered state.  In REM you can invert these so it does a good job of catching it as quickly as possible but depending on how the relay board is powered it can cycle to its final destination.  So with power applied to the relay board and no power applied to the pi, all of the pins would be low.  This means that if it were a low level trigger then the state would have the relay engaged.  On the flip side of that with no power applied to the relay board then the relays are disengaged because there is nothing powering the relay solenoid.  The reason why they make those things is because the circuit is much simpler because no power is required to hold the engaged position. 

Wondering if that modbus relay has the same problem...

**Reply by orrious** [reply]
That's a Rpi issue and why I won't use the GPIOs on the RPi either.  I'm using an arduino Mega currently because of this issue.  I would be surprised if the RS485 relay board suffers from this problem.

**Reply by orrious** [reply]
> `@orrious: I would not drive one relay with another `
> 
> It’s done all the time, Inquiring minds want to know why not?

Closed contact relays are mechanical devices and have a Mean Time to Failure and therefor only be used when needed.  When do you need to a relay?  When you want to control a lot of current / voltage with a little bit.  Using one relay to trigger another creates an unneeded additional point of failure that will happen eventually.  

The only relays in my entire system that aren't solid state are the ones that control the Intellivalves.  They do not draw enough current to bias the solid state relay.  I can't wait till we can control them via RS485, another selfish reason I want RS485 in that enclosure.

**Reply by rstrouse** [reply]
While it is true that SS relays will outlast electromechanical models the MTBF on them given our duty cycles is pretty long.  Probably longer than most parts on your pool.  Oddly enough turning a relay off is the what causes the damage.  SS relays are really quite nice but their Achilles heel is thermal management and good designs typically incorporate large heat sinks.  The ones that I saw in @orrious pics are clearly top quality relays.

You can also perform switching with them that is crazy fast.  Some models can be switched switched in the 50 microsecond range.  SSRs also require very little on current.  This is why many of the relays you see out there are low level triggers which means that the pin is high when they are off.  This is not necessarily a problem with the pi rather it is an assumption that there will be more available current when the relay is in the NO position.  This is design choice by the manufacturer of the relay and often amounts to a little bit of cost savings since the isolation circuits require fewer components.

The reason why this appears to be a problem on the pi vs arduino is that the pi uses a microprocessor vs a microcontroller and by the time the pi it gets around to initializing its GPIO, power is already available to the microcontroller on the relay board.  So it engages the relay during boot.

Simply using a high level I/O trigger instead of a low level trigger is the fix to the problem.  In the wild will find relays with triggers that are biased high and relays that biased low.  I have seen some folks that pull the I/O low with an external resistor to reverse the trigger.  NO then becomes NC.

I don't think I have ever seen an SSR with a low level trigger since the current requirements for the trigger are not even a concern.  Kinda tells me that the primary driver for low level relays is saving fractional cents for a circuit to pull the trigger low.

**Reply by orrious** [reply]
@rstrouse all of my relays are HI level triggered.  During initial work using RPi GPIO and 3.3v - 5v opto isolators to control the SSRs, I noticed when the RPi would reboot GPIOs would go HI and turn on pumps and valves until it finished boot.  I was worried that the SD card would fail and the RPi would not fully boot; therefore leaving the GPIOs in an unmanaged / HI state and pumps potentially dead headed.

**Reply by rstrouse** [reply]
The pi GPIO does not go high it floats until the overlays load.

**Reply by orrious** [reply]
Just looked at the level shifter I was using.  It's using pull up resistors, so if the GPIOs float until initialized, the level shifter I used would have pulled the outputs HI.  Without using a MCP23017 like I2C board, any recommendations for reliable 5v GPIO output using the RPi native GPIOs?

**Reply by rstrouse** [reply]
There are a number of pi hats that use gpio with a high level trigger.  However the number of cool relay boards that I used to see have been greatly reduced.  Honestly I really like the Sequent8 hats but I realize this presents challenges for you.  Might be the only way to get a perfect solution is to build your own.  I don’t even see the number of industrial options I used to see.

**Reply by orrious** [reply]
I've essentially already built my own board using the Arduino Mega ([pool-i2c-mega](https://github.com/orrious/pool-i2c-mega)) and am trying to not maintain it as a one-off.  It works well, I just don't want to be the only one able to support it and thus my driving motivation to use a more widely available solution.

There doesn't appear to be a platform that is really focused on the pool automation space.  Atlas is starting to dabble in that direction, but it's ESP32 based and won't fit the needs of this project.  (but look at the price, they clearly think there is demand.)  Sequent8 is focused on industrial controls and building automation, not pools.

That being said, I would be interested in collecting our requirements, as a community.  Agree on what we would like for a gen 1 platform and approach Sequent8.  If they aren't interested, I will happily help design the hardware, write the code for the MPU, and open source it all.  We can easily have them built and assembled as prototypes, 5+ at a time.

**Reply by tagyoureit** [reply]
We built [REM](https://github.com/rstrouse/relayEquipmentManager) to automate the chemical dosing, control relays, interact with all different kinds of boards, etc.  And we do use/recommend/prefer the Sequent cards as being the best out there.

**Reply by orrious** [reply]
@tagyoureit I assure you I would prefer to buy something off the shelf and I like Sequent quite a bit, however none of the Sequent cards fit my needs.  I need 20+ 5v digital outputs to run my own SSRs.

I just order this to play with.  Lets see if I ever get it:
https://www.aliexpress.com/item/3256803335438797.html?gatewayAdapt=4itemAdapt

**Comment by orrious** [comment]
I got my MEGA-BAS recently.  Ironically, by default the RS485 is in slave mode and you can fully control the MEGA-BAS via modbus without the need for a RPi local.  I think this provides further credence to the request for modbus support in REM towards the goal of a Pi-less pad.

**Reply by rstrouse** [reply]
But you cannot put those commands on an RS485 bus that also contains an RS485 pump, chlorinator, heater, and maybe valves one day.  They are not strictly modbus.  It would be like dogs and cats living together.  One of them is going to lose when the food runs out.  For this application the only thing on that bus would be the MEGA-BAS and the pumps, chlorinators, heaters, valves, and OCP if it exists would be chatting it up on a whole other bus.

**Reply by orrious** [reply]
Completely agree, this would need it's own RS485 modbus bus, but so would supporting a RS485 modbus relay board.  REM already supports the concept of multiple busses per type of protocol.

**Reply by rstrouse** [reply]
One of the main ideas of that port on the MEGA-BAS is to leverage it for RS485 comms to the equipment.  If I had more ambition maybe I would create a REM version to run on an ESP and interface that with microservices.  While there is a lot of development there, unfortunately, there are so many half baked libraries on that platform.  Lots of refactoring to get to a final result when you think its the official library.

I guess if you live on them you know what is what.  However, if you dabble there is nothing stopping you from downloading one of a hundred socket implementations that sends strange opcode bytes when you emit on the socket but it translates incoming beautifully.  Rabbit holes like that just eat up precious dev time. (sounds like somebody has been burned)

All that being said though much of this is driven by what equipment is available.  The availability of I2c driven peripheral devices for switching, sensing, and controlling is what got us here.  As for remote RS485 for equipment there are quite a few inexpensive socket based RS485 servers that can be installed at the pad that put the comms on category cable.

Lets see where we get up to with the PCF chips.  That latest order gets to 48 GPIO with the three breakouts and a limit of 112 with hardware addressing.
